src/metadata/mds.rs

```diff
@@ -0,0 +1,729 @@
+
+            if let Some(wal_files) = &scope.wal_ids {
+                let mut query_builder =
+                    QueryBuilder::new("INSERT INTO scope_wal_files (file_id, scope_id)");
+                query_builder.push_values(wal_files.iter(), |mut b, wal_file| {
+                    b.push_bind(wal_file.to_string()).push_bind(scope_id);
+                });
+                let _ = query_builder.build().execute(&mut *tx).await?;
+            }
+
+            sqlx::query(
+                r#"
+                INSERT INTO version_snapshot_leveled_scopes 
+                (version_snapshot_id, row_index, col_index, scope_id)
+                VALUES ($1, $2, $3, $4)
+            "#,
+            )
+            .bind(version_snapshot_id)
+            .bind(row_idx as i32)
+            .bind(col_idx as i32)
+            .bind(scope_id)
+            .execute(&mut *tx)
+            .await?;
+        }
+    }
+    tx.commit().await?;
+
+    Ok(())
+}
+
+impl<R> MetadataManifestStorage<R>
+where
+    R: Record,
+{
+    pub async fn new(
+        conn_url: impl Into<String>,
+        clean_sender: Sender<CleanTag>,
+        option: Arc<DbOption>,
+    ) -> Result<Self, MetadataServiceError> {
+        let pool = PgPoolOptions::new()
+            .idle_timeout(Some(Duration::from_secs(10 * 60)))
+            .max_connections(10)
+            .connect(&conn_url.into())
+            .await?;
+
+        let mut tx = pool.begin().await?;
+        let version_snapshot = sqlx::query_as::<_, VersionSnapshot>(
+            r#"
+                SELECT id, timestamp, log_length
+                FROM version_snapshots 
+                ORDER BY timestamp DESC 
+                LIMIT 1
+            "#,
+        )
+        .fetch_optional(&mut *tx)
+        .await?;
+        tx.commit().await?;
+
+        let mut timestamp = Arc::new(AtomicU32::default());
+        if let Some(version_snapshot) = version_snapshot {
+            timestamp = Arc::new((version_snapshot.timestamp as u32).into());
+        } else {
+            let new_version = Arc::new(Version::<R>::new_full(
+                Timestamp::from(0),
+                0,
+                [const { Vec::new() }; MAX_LEVEL],
+                option.clone(),
+                clean_sender.clone(),
+                Arc::new(AtomicU32::new(0)),
+            ));
+            let mut tx = pool.begin().await?;
+            write_version_snapshot(new_version, tx).await?;
+        }
+
+        Ok(Self {
+            pool,
+            clean_sender,
+            timestamp,
+            option,
+            _marker: PhantomData,
+        })
+    }
+
+    async fn read_current(&self) -> Result<Option<VersionRef<R>>, MetadataServiceError> {
+        let mut tx = self.pool().begin().await?;
+
+        let version_snapshot = sqlx::query_as::<_, VersionSnapshot>(
+            r#"
+                SELECT id, timestamp, log_length
+                FROM version_snapshots 
+                ORDER BY created_at DESC 
```

Comment:
the ordering column is different with https://github.com/tonbo-io/tonbo/pull/481/files#diff-f2ac089b8c9efebe62753973e425342ec1abbb4a4dee365daa6babca6e8a35deR140, it would be clear if we always use the same column to get/set the current timestamp